Update instruction macros to use .insn directive

pranav-prakash opened the pull request:
The prior solution relied on deprecated register storage class specifiers.
Note that .insn directives require a sufficiently recent version of binutils.

Addresses issue #3.

DCO 1.1 Signed-off-by: Pranav Prakash <pranav-prakash@users.noreply.github.com>

pranav-prakash committed:
Update instruction macros to use .insn directive

The prior solution relied on deprecated register storage class specifiers.
Note that .insn directives require a sufficiently recent version of binutils.

Addresses issue #3.

DCO 1.1 Signed-off-by: Pranav Prakash <pranav-prakash@users.noreply.github.com>

seldridge commented on src/xcustom.h:
```diff
@@ -19,62 +19,32 @@
-  (rs2                  << (7+5+3+5)) |         \
-  (EXTRACT(funct, 7, 0) << (7+5+3+5+5))
-
-#define ROCC_INSTRUCTION_RAW_R_R_R(x, rd, rs1, rs2, funct)      \
-  .word XCUSTOM(x, ## rd, ## rs1, ## rs2, funct)
```
Can the PR leave this in? 

I still use this for some straight riscv-test-style assembly language tests where I'm actually writing assembly as opposed to inline assembly stuck into a C program.

pranav-prakash replied:
Seems like this has been addressed this in the PR you PR'd against my PR.

seldridge reviewed (changes requested):
This is awesome. Thanks for taking the time to do what I didn't do for this.

I was playing around with this locally and had some modifications that I'll PR against this PR now.

seldridge committed:
Fix xd, use CUSTOM_* opcode names, re-add RAW

- Fix incorrect setting of xd value when rd is expected.
- Switch to using CUSTOM_* opcode names instead of hard-coding it. Do
  this with a two-level macro expansion, CAT, so users can pass in
  macros as the x parameter.
- Re-add the removed *_RAW_* instruction variants as these are useful
  for bare-metal assembly language tests (like riscv-tests).

Signed-off-by: Schuyler Eldridge <schuyler.eldridge@ibm.com>

seldridge committed:
Bump copyright

Signed-off-by: Schuyler Eldridge <schuyler.eldridge@ibm.com>

seldridge committed:
Add Scaladoc-style documentation

Signed-off-by: Schuyler Eldridge <schuyler.eldridge@ibm.com>

pranav-prakash committed:
Fix style and update documentation for C macros

* Format based on Google style guide
* Update documentation for C/C++ macros to more accurately reflect that `rd`, `rs1`, and `rs2` are C variables passed in and captured as operands (i.e. they are not directly registers as is the case for the bare-metal macro).

Signed-off-by: Pranav Prakash <pranav-prakash@users.noreply.github.com>

pranav-prakash committed:
Merge pull request #1 from IBM/pr-5-review

PR #5 Review.

pranav-prakash commented:
I merged the changes you PR'd against my fork so I believe everything should be good to go.

I've been testing the C macros against proxy kernel since I'm using this for gemmini (née systolic) and it works fine there. And you mentioned you tested out bare metal so that is hopefully thorough enough.

seldridge commented:
This is great! Thanks for getting this together.
